File: Lab_7/Codes/FFT.m
% List of image files to process
imageFiles = {'moody.jpg', 'baby2.jpg', 'flower.jpg'}; % Add more image file names if needed
numImages = length(imageFiles);
% Create a figure to display the results
figure;
% Loop through each image file
for i = 1:numImages
    % Read the image
    originalImage = imread(imageFiles{i});
    
    % Check if the image has multiple color channels
    if size(originalImage, 3) == 3
        % Perform 2D FFT for each channel individually
        fftImageR = fft2(double(originalImage(:,:,1)));
        fftImageG = fft2(double(originalImage(:,:,2)));
        fftImageB = fft2(double(originalImage(:,:,3)));
        % Shift the zero frequency component to the center for each channel
        fftImageShiftedR = fftshift(fftImageR);
        fftImageShiftedG = fftshift(fftImageG);
        fftImageShiftedB = fftshift(fftImageB);
        % Compute the magnitude spectrum for each channel
        magnitudeSpectrumR = log(1 + abs(fftImageShiftedR)); % Red channel
        magnitudeSpectrumG = log(1 + abs(fftImageShiftedG)); % Green channel
        magnitudeSpectrumB = log(1 + abs(fftImageShiftedB)); % Blue channel
        % Compute the average magnitude spectrum for RGB channels
        magnitudeSpectrumAvg = (magnitudeSpectrumR + magnitudeSpectrumG + magnitudeSpectrumB) / 3;
        
    else
        % Perform FFT on the grayscale image
        fftImage = fft2(double(originalImage));
        fftImageShifted = fftshift(fftImage);
        magnitudeSpectrumAvg = log(1 + abs(fftImageShifted));
    end
    % Display the original image
    subplot(numImages, 2, 2*i-1);
    imshow(originalImage);
    title(['Original Image ', num2str(i)]);
    
    % Display the average magnitude spectrum (log scale)
    subplot(numImages, 2, 2*i);
    imshow(magnitudeSpectrumAvg, []);
    title(['Magnitude Spectrum ', num2str(i)]);
end

File: Lab_7/Codes/FFT.py
import cv2
import numpy as np
import matplotlib.pyplot as plt

# List of image files to process
image_files = ['moody.jpg', 'baby2.jpg', 'flower.jpg']

# Create a figure to hold all the subplots
plt.figure(figsize=(15, 10))

# Loop through each image file
for i, image_file in enumerate(image_files):
    # Load the image
    original_image = cv2.imread(image_file)
    original_image_rgb = cv2.cvtColor(original_image, cv2.COLOR_BGR2RGB)

    # Check if it's a color image
    if original_image_rgb.shape[2] == 3:
        # Perform FFT for each channel
        fft_images = [np.fft.fft2(original_image_rgb[:, :, channel]) for channel in range(3)]
        fft_images_shifted = [np.fft.fftshift(fft_image) for fft_image in fft_images]
        magnitude_spectrum = [np.log(1 + np.abs(fft_shifted)) for fft_shifted in fft_images_shifted]
        magnitude_spectrum_avg = sum(magnitude_spectrum) / 3
    else:
        fft_image = np.fft.fft2(original_image_rgb)
        fft_image_shifted = np.fft.fftshift(fft_image)
        magnitude_spectrum_avg = np.log(1 + np.abs(fft_image_shifted))

    # Plot the original image
    plt.subplot(len(image_files), 2, 2 * i + 1)  # Row: len(image_files), Col: 2, Pos: i*2+1
    plt.imshow(original_image_rgb)
    plt.title(f'Original Image {i+1}')
    plt.axis('off')  # Hide axis

    # Plot the magnitude spectrum
    plt.subplot(len(image_files), 2, 2 * i + 2)  # Row: len(image_files), Col: 2, Pos: i*2+2
    plt.imshow(magnitude_spectrum_avg, cmap='gray')
    plt.title(f'Magnitude Spectrum {i+1}')
    plt.axis('off')  # Hide axis

# Adjust layout to prevent overlapping
plt.tight_layout()
plt.show()
